coccinelle: fix the incorrect transformation/isomorphisms

mrc0mmand opened the pull request:
To avoid the incorrectly formatted code after the *equals-null* transformation I introduced a custom isomorphism file which worked around the issue. However, by misreading the documentation, this isomorphism file turned off all standard isomorphisms and also masked several other issues.

After correcting the respective transformation and dropping the isomorphism file, some new, previously masked "issues" popped out (see the last commit).

Unfortunately, by re-enabling all the standard isomorphisms, the complexity of the *in_set* and *not_in_set* transformations rose exponentially (with 8 jobs in parallel it tried to eat almost 24G of memory). so by limiting it to four parallel jobs I managed to go through this transformation as well, but even then it took around 30 minutes for each of them to complete. I'll try to look into this as well as on one false positive which popped out.

/cc @poettering, hopefully this fixes your issue

poettering commented on src/basic/log.c:
```diff
@@ -353,7 +353,7 @@ void log_forget_fds(void) {
 
 void log_set_max_level_realm(LogRealm realm, int level) {
-        assert((level & LOG_PRIMASK) == level);
+        assert(FLAGS_SET(LOG_PRIMASK, level));
```
hmm, this is a weird one. This check was added here to validate that the specifier level argument has no bits set but the level. With the change now it reads a bit weirdly, as in "check that `LOG_PRIMASK` has all bits in `level` set", which is still correct, but a weird way to express what we want to check here...

maybe the fix here is to change the expressio to:

```c
assert(l(evel & ~LOG_PRIMASK) == 0);
```
which is equivalent, but I think more explicitly as it denotes exactly: all other bits than the ones from LOG_PRIMASK should be zero.

poettering commented on src/basic/utf8.c:
```diff
@@ -42,7 +42,7 @@ bool unichar_is_valid(char32_t ch) {
         if ((ch >= 0xFDD0) && (ch <= 0xFDEF)) /* Reserved */
                 return false;
-        if ((ch & 0xFFFE) == 0xFFFE) /* BOM (Byte Order Mark) */
+        if (FLAGS_SET(ch, 0xFFFE)) /* BOM (Byte Order Mark) */
```
hmm, this is a bit misleading I think. `FLAGS_SET()` suggests in its name that we check flags here, but that's not the case. we check a non-flag value here, masking off some upper and lower bits... I think this should stay as it was.

poettering commented on src/core/manager.h:
```diff
@@ -136,7 +136,7 @@ typedef enum ManagerTestRunFlags {
 } ManagerTestRunFlags;
 
-assert_cc((MANAGER_TEST_FULL & UINT8_MAX) == MANAGER_TEST_FULL);
+assert_cc(FLAGS_SET(UINT8_MAX, MANAGER_TEST_FULL));
```
hmm, so this check was supposed to check that MANAGER_TEST_FULL fits in 8 bits. The FLAGS_SET() epxression still checks that, but again is a bit misleading, since there are not flags to be checked here...

i tmight make sense to simply rewrite this test as: MANAGER_TEST_FULL <= UINT8_MAX, to make it clearer this is about range

poettering reviewed (changes requested):
some minor issues. The FLAGS_SET cocci script is a bit too greedy i guess. it's about checking flags but not all equivalent bitwise expressions are about checking flags...

mrc0mmand commented:
> some minor issues. The FLAGS_SET cocci script is a bit too greedy i guess. it's about checking flags but not all equivalent bitwise expressions are about checking flags...

Thanks a lot for the comments. I dropped the referenced changes and added the respective files/functions into the *exclusion list* in the *flags-set* transformation.

mrc0mmand commented:
Just a note: I wouldn't recommend running the `coccinelle/run-coccinelle.sh` script for now unless you have a really beefy machine (or just temporarily remove the *in_set* and *not_in_set* transformation files) otherwise it's going to swap _a lot_ and run for hours without any meaningful results.

I've started with some light profiling, and there is something really expensive in these transformations. For example:

 ### With standard isomorphisms (`/usr/lib64/coccinelle/standard.iso`)
```
$ spatch --profile --profile-iso --sp-file coccinelle/in_set.cocci src/core/execute.c 
init_defs_builtins: /usr/lib64/coccinelle/standard.h
HANDLING: src/core/execute.c
Note: processing took    54.3s: src/core/execute.c
---------------------
profiling result
---------------------
full_engine                              :     54.253 sec          1 count
mysat                                    :     53.585 sec        617 count
pre_engine                               :      3.156 sec          1 count
asttoctl2                                :      2.306 sec          1 count
parse cocci                              :      0.843 sec          1 count
```

### Without standard isomorphisms
```
$ spatch --iso /dev/null --profile --profile-iso --sp-file coccinelle/in_set.cocci src/core/execute.c 
init_defs_builtins: /usr/lib64/coccinelle/standard.h
warning: Can't find default iso file: /dev/null
HANDLING: src/core/execute.c
---------------------
profiling result
---------------------
full_engine                              :      0.643 sec          1 count
mysat                                    :      0.028 sec        617 count
pre_engine                               :      0.009 sec          1 count
parse cocci                              :      0.009 sec          1 count
asttoctl2                                :      0.000 sec          1 count
```

mrc0mmand commented:
Ah, I see where the *issue* lies. Thanks to me, the default isomorphisms were disabled, so an SmPL expression `e == n0 || e == n1 || e == n2 || e == n3` always had to have an expression first and then a constant. However, in the current state, the various isomorphisms are able to do permutations on such expressions, thus:

this expression:
```
e == n0 || e == n1 || e == n2 || e == n3
```
is expanded to this set of expressions:

```
|
e == n0 || e == n1 || e == n2 || e == n3
|
e == n0 || e == n1 || e == n2 || n3 == e
|
e == n0 || e == n1 || n2 == e || e == n3
|
e == n0 || e == n1 || n2 == e || n3 == e
|
e == n0 || n1 == e || e == n2 || e == n3
|
e == n0 || n1 == e || e == n2 || n3 == e
|
e == n0 || n1 == e || n2 == e || e == n3
|
e == n0 || n1 == e || n2 == e || n3 == e
|
n0 == e || e == n1 || e == n2 || e == n3
|
n0 == e || e == n1 || e == n2 || n3 == e
|
n0 == e || e == n1 || n2 == e || e == n3
|
n0 == e || e == n1 || n2 == e || n3 == e
|
n0 == e || n1 == e || e == n2 || e == n3
|
n0 == e || n1 == e || e == n2 || n3 == e
|
n0 == e || n1 == e || n2 == e || e == n3
|
n0 == e || n1 == e || n2 == e || n3 == e
```

As the expressions in our in_set/not_in_set transformations go up to nine *subexpressions* per expression, this generates a huge set which takes a lot of memory (1.8G on my machine) and takes a while to match against a file (like the 54s against `src/core/execute.c`). I'll try to dig a bit deeper, if this can be optimized somehow.

mrc0mmand commented:
So, I did some more reading, testing, and benchmarking and the results are as follows:

The transformations can be either written separately or in a disjunction. The former creates a map of each transformation separately and attempts to match it against statements in a file. The later one creates a map of the whole disjunction statement and again, tries to match it (that's the current state). The second approach is in this specific case slower and much more resource demanding.

I did some benchmarking on my desktop (Ryzen 5 3600 6 core + HT @ 4GHz, 16GB RAM) for comparison:

1) **split**, 9 expressions, 12 threads, ~800M per thread ~= 9.6G of memory
```
real	21m30.092s
user	236m51.564s
sys	1m7.206s
```

2) **split**, 8 expressions, 12 threads, ~300M per thread ~= 3.6G of memory
```
real	7m24.369s
user	81m21.195s
sys	0m24.098s
```

3) **disjunction**, 8 expressions, 12 threads, ~550M per thread ~= 6.6G of memory
```
real	14m15.858s
user	157m24.670s
sys	0m48.582s
```

4) **split**, 7 expressions, 12 threads, ~200M per thread ~= 2.4G of memory
```
real	2m19.783s
user	25m38.592s
sys	0m11.070s
```

I also tried the disjunction version with the above settings, but it has eaten through all the 16 GB of memory and 8 GB of swap, so after 4.5 hours of trashing I decided to kill it :-)

Given the benchmark results, I'd rewrite both *in_set* and *not_in_set* transformations and top them at 7 *subexpressions* instead of the current 9. In my opinion it's still going to catch most of the cases and the 7+ cases would be still pretty obvious during reviews to catch manually.

@poettering what do you think?

keszybz commented on man/hwdb-usb-device.c:
```diff
@@ -9,7 +9,7 @@ int print_usb_properties(uint16_t vid, uint16_t pid) {
 
   /* Match this USB vendor and product ID combination */
-  snprintf(match, sizeof match, "usb:v%04Xp%04X", vid, pid);
+  xsprintf(match, "usb:v%04Xp%04X", vid, pid);
```
No, this is in man, we don't expose xsprintf to users.

mrc0mmand replied:
Thanks, fixed. I also added a rule to skip the `xsprintf` trasnformation on `man/*`.

keszybz commented:
> Given the benchmark results, I'd rewrite both _in_set_ and _not_in_set_ transformations and top them at 8 expressions instead of the current 10. In my opinion it's still going to catch most of the cases and the 8+ cases would be still pretty obvious during reviews to catch manually.

I'd limit it even more. Reasonable behaviour is more important than catching those long cases which are obvious anyway.

mrc0mmand commented:
> > Given the benchmark results, I'd rewrite both _in_set_ and _not_in_set_ transformations and top them at 8 expressions instead of the current 10. In my opinion it's still going to catch most of the cases and the 8+ cases would be still pretty obvious during reviews to catch manually.
> 
> I'd limit it even more. Reasonable behaviour is more important than catching those long cases which are obvious anyway.

Definitely aggree. After a bit more "benchmarking" on my work laptop (which will be soon 6 years old) I settled the limit at 6 expressions, which should give a good balance between resource usage & helpfulness.

The `strjoin` commit just "resolves" one false positive, so now the output of coccinelle/spatch is clean of all false positives, which should make it more useful.
